libbeat/monitoring/report/elasticsearch/client.go

```diff
@@ -245,25 +251,38 @@ func getMonitoringIndexName() string {
 	return fmt.Sprintf(".monitoring-beats-%v-%s", version, date)
 }
 
-func logBulkFailures(result esout.BulkResult, events []report.Event) {
-	reader := esout.NewJSONReader(result)
-	err := esout.BulkReadToItems(reader)
-	if err != nil {
-		logp.Err("failed to parse monitoring bulk items: %v", err)
+func logBulkFailures(result eslegclient.BulkResult, events []report.Event, logger *logp.Logger) {
+	var response struct {
+		Items []map[string]map[string]interface{} `json:"items"`
+	}
+
+	if err := json.Unmarshal(result, &response); err != nil {
```

Comment:
This change might be somewhat heavy, as it allocates quite a many object now per response. The old reader was used to step through the responses without requiring extra allocations. `(json.Decoder).More/Token` can achieve something similar. Would be interesting to see some benchmarks how this change affects memory/CPU usage with filebeat->ES.

Reply:
I made this change in response to your comment here: https://github.com/elastic/beats/pull/16150#discussion_r379008004. In particular, I was thinking that since monitoring's "bulk" request really only contains 1 event/doc, the corresponding bulk response will also only contain one result. So I thought it was fair to trade the "heaviness" for using the regular bulk API (i.e. not using the streaming parser). It means exporting fewer symbols and keeping the streaming parser just within the ES output package.

Reply:
Correct me if I'm wrong, but the output.elasticsearch package also goes through this code path, right?

For monitoring performance is indeed not my concern. It's like one API call with one event every 30s or so. But we should not use `json.Unmarshal` on the Items response via the output, which would be directly in the hot code path.

Reply:
I confirmed that this function is only called from the monitoring Elasticsearch reporter code path. It is called from here: https://github.com/elastic/beats/pull/16150/files#diff-50c005cc3d307398fcfcc0c1920ff0e4R243 (line 243, just a few lines above this one).

As for how the Elasticsearch output's bulk response get's processed, that's taking a different code path, starting over here: https://github.com/elastic/beats/pull/16150/files#diff-524ca932819c545af8f54fd09b9c2c5eR227. The function being called there, `bulkCollectPublishFails`, is defined over here: https://github.com/elastic/beats/pull/16150/files#diff-524ca932819c545af8f54fd09b9c2c5eR345. You can see it is using the streaming JSON parser.

I think the reason you didn't see this was because that file, `libbeat/outputs/elasticsearch/client.go`, was collapsed by GitHub due to the diff being too large.

Reply:
Haha, you removed so many lines from this file, I thought the file got deleted :)

Thanks for the confirmation. Would it make sense to move the JSON scanner into the output/elasticsearch package? Bulk request support in go-elasticsearch is still in progress, I presume we can throw away the scanner in the future.

Reply:
> Would it make sense to move the JSON scanner into the output/elasticsearch package? Bulk request support in go-elasticsearch is still in progress, I presume we can throw away the scanner in the future.

Yes! Done in f74cc1ac305899a5289779a74444fa8d27b27590.